agent/cache-types/connect_ca_leaf.go

```diff
@@ -48,11 +48,15 @@ func (c *ConnectCALeaf) Fetch(opts cache.FetchOptions, req cache.Request) (cache
 	newRootCACh := make(chan error, 1)
 	go c.waitNewRootCA(reqReal.Datacenter, newRootCACh, opts.Timeout)
 
+	// Generate a cache key to lookup/store the cert. We MUST generate a new cert
+	// per token used to ensure revocation by ACL token is robust.
+	issuedKey := fmt.Sprintf("%s/%s", reqReal.Service, reqReal.Token)
```

Comment:
This may be okay but I would probably prefer to avoid concatenation as it has the very minute potential to cause other collisions and becomes an attack vector for gaining access to certificates. It makes using '/' in service names or in the tokens a little unsafe (neither of which is good idea to begin with). One thing I learned the other day is that you can use structs as map keys.

```
type issueKey struct {
   Service string
   Token string
}

issuedKey := issueKey{reqReal.Service, reqReal.Token}
lastCert := c.issuedCerts[issuedKey]
```

Alternatively you could do an ordered hash over the two parts similar to how what hashstructure would do.

In the end this probably doesn't matter. Its just something I started considering while working on ACLs.

Reply:
Reasonable, I've been getting twitchy about secrets being used in keys and so being stuff randomly in memory we don't think of as "sensitive" to. This was just following the pattern from Cache.

I'll change to a hash I think.